src/minisphere/event_loop.c

```diff
@@ -71,6 +126,65 @@ events_run_main_loop(void)
 	}
 }
 
+void
+events_tick(int api_version, bool clear_screen, int framerate)
+{
+	size_t              bytes_read;
+	struct socket_read* sr;
+
+	int i, len;
+
+	sphere_heartbeat(true, api_version);
+
+	if (!screen_skipping_frame(g_screen)) {
+		if (!dispatch_run(JOB_ON_RENDER))
+			return;
+	}
+	screen_flip(g_screen, framerate, clear_screen);
+	if (api_version >= 2)
+		image_set_scissor(screen_backbuffer(g_screen), screen_bounds(g_screen));
+
+	if (!dispatch_run(JOB_ON_UPDATE))
+		return;
+
+	// handle asynchronous socket reads
+	for (i = 0, len = vector_len(s_socket_reads); i < len; ++i) {
+		sr = vector_get(s_socket_reads, i);
+		bytes_read = socket_read(sr->socket, sr->ptr, sr->bytes_left);
```

Comment:
I just want to point out that this is actually somewhat suspect although it may not look like it at first; it works right now only because data is buffered on the main thread and requires regular calls to `sockets_update()` to feed the receive buffer.  So FIFO order is guaranteed by the `socket_read` being greedy, but we won't be able to rely on that if sockets are ever moved off-thread.

The correct thing to do here if we were to move to off-thread sockets would be, if there are multiple read requests for the same socket, on each round of the event loop to to only service the first one in the queue until it's full and ignore subsequent ones.  Otherwise async reads could accidentally "steal" data from an each other...